feat: manage tokens

cbadawi opened the pull request.

cbadawi commented on src/app/features/asset-list/bitcoin/brc20-token-asset-list/brc20-token-asset-list.tsx:
```diff
@@ -22,11 +23,27 @@ interface Brc20TokenAssetDetails {
   const navigate = useNavigate();
   const { balance, query: result } = useCurrentBtcCryptoAssetBalanceNativeSegwit();
 
+  if (!tokens.length) return null;
```
inadvertently fixed issue #5469

the stack was creating the extra div even where there were no tokens

pete-watters replied:
Nice! How about adding that fix in a separate PR so we can get it released sooner?

kyranjamie commented on src/app/store/manage-tokens/manage-tokens.hooks.ts:
```diff
@@ -0,0 +1,17 @@
+import { addToken } from './manage-tokens.slice';
+
+// loads tokens to store if it doesnt exist & sets enabled as false
+function useInitTokensAsDisabled({ ids }: { ids: string[] }) {
```
I think there might be a simpler implementation for the store, that doesn't pre-setting of state. 

Conceptually, we have two sets of binary states. Our defaults, supported/unsupported, and the new ones you've added, a user set enabled/disabled.

Suggest implementing this in such a way where, by default, the `manageTokens` slice is empty, as users have made no state changes. An `undefined` for a look up of the token in the manageTokens slice means the user as set no preference, fallback to default. 

Only when modifying a rule, would we then save the state. I think this way we could avoid the initial "set all the state" action.

kyranjamie commented on src/app/features/manage-tokens/manage-tokens.tsx:
```diff
@@ -0,0 +1,111 @@
+  onClose(): void;
+}
+
+export type TokenProtocol = 'stacks' | 'bitcoin';
```
This already exists as `SupportedBlockchains`, can import that instead.

pete-watters replied:
Can you update this @cbadawi ?

kyranjamie commented on src/app/features/asset-list/stacks/stx20-token-asset-list/stx20-token-asset-list.tsx:
```diff
@@ -10,8 +12,16 @@ interface Stx20TokenAssetDetails {
 }
-export function Stx20TokenAssetList({ tokens }: Stx20TokenAssetListProps) {
+export function Stx20TokenAssetList({ tokens, countSetter }: Stx20TokenAssetListProps) {
+  useEffect(() => {
+    if (countSetter) {
+      countSetter(tokens.length);
+    }
+  }, [tokens, countSetter]);
```
Why do we need this `countSetter`? Can't we infer the value another way to avoid having to do this?

The parent passes in `tokens`, so it knows the `length` already.

cbadawi replied:
Handled them in the bitcoin & stacks list components. happy to take feedback on a cleaner way to have the manage-tokens component know if the children have any tokens.

```
{
                useEffect(() => {
                  if (hasTokenSetter && tokens.length) hasTokenSetter(true);
                });
                return <Stx20TokenAssetList tokens={tokens} />;
              }}
```

cbadawi replied:
above comment is no longer up-to-date. I had to re-add them to the [token]-asset-lists components above.

```
  useEffect(() => {
    if (hasTokenSetter && tokens.length) hasTokenSetter(true);
  }, [tokens.length, hasTokenSetter]);
```

The state update needs to happen in a use-effect. hence cant be done in a higher component.
```
sip10-token-asset-list.tsx:28 Warning: Cannot update a component (`ManageTokensDialog`) while rendering a different component (`Sip10TokenAssetList`). To locate the bad setState() call inside `Sip10TokenAssetList`, follow the stack trace as described in 
```
https://stackoverflow.com/questions/62336340/cannot-update-a-component-while-rendering-a-different-component-warning

happy to get feedback here on a cleaner approach. we can potentially use redux instead of lifting the state up.

@kyranjamie @pete-watters

kyranjamie reviewed (commented):
Thanks for the PR @cbadawi

Please be patient while the team reviews. @fbwoolf and @alter-eggo should also review this, as they've worked on this part of the codebase recently. We'd also like to get designer input on this @fabric-8 @mica000

markmhendrickson commented:
@fabric-8 @mica000 Mind reviewing and providing input on the UX generally?

See https://x.com/anononchain/status/1804308553982673022

fabric-8 reviewed (commented):
Hey @cbadawi thanks a lot for working on this!

Here are some suggestions from the design side:

<img width="1280" alt="hide tokens" src="https://github.com/leather-io/extension/assets/4112698/3e7d82c0-a002-40d2-978a-7d21e7de7b3b">

- Can we simply replace "show more/less" with "Manage tokens"? This would reduce the amount of related elements we show top level while still allowing for an easy entry point in case people wonder about missing tokens.
- As of the token visibility list itself, I think we don't need to split tokens by the chain since the vast majority of spam tokens happen on STX anyways, so just one tabless list should suffice. Furthermore, let's adjust the toggle a bit to look more coherent with the rest of the app.

[You can inspect the design mockups here](https://www.figma.com/design/FlmF7ropcijOb7iVSDqNn0/%F0%9F%9F%A7-Web-App-Explorations-2024-H1?node-id=10337-11339&t=qfMMR7tJ1m3gRr9F-1) — Feel free to reach out in case you have questions or feedback!

cbadawi commented:
sorry for being late here, I mostly just have free time on weekends. The fixes are done apart from these design questions :

- I could not change the light/dark theme on the figma files (maybe lack access?). Could I get screenshots of the toggle in dark mode please?

- silly question ... but I couldn't find the ">" icon in `/@leather-wallet/ui/dist-all/src/icons` the expand icon is the closest thing to it. Am i looking at the wrong icon folder? should I just use ">" ?

@kyranjamie @fabric-8

fbwoolf commented:
> * silly question ... but I couldn't find the ">" icon in `/@leather-wallet/ui/dist-all/src/icons` the expand icon is the closest thing to it. Am i looking at the wrong icon folder? should I just use ">" ?
> 
Adding the icon you need here: https://github.com/leather-io/mono/pull/254 ...we will need to release the change for you to install here.

fabric-8 commented:
@cbadawi Besides the icon which should be available by now once you pull in the latest changes, I've added a [dark mode variant](https://www.figma.com/design/FlmF7ropcijOb7iVSDqNn0/%F0%9F%9F%A7-Web-App-Explorations-2024-H1?node-id=10821-83107&t=v68VlFd2rongNVUT-1) as well — If you just use the color tokens as in the designs all of that will take care of itself :)

cbadawi commented:
thanks @fbwoolf  @fabric-8 , just noticed the color tokens & applied the changes. 
couple of minor issues though : 
- could not make the color tokens work on the border and had to hardcode the colors
- the chevron right icon has 2 chevrons. ">>"

https://github.com/leather-io/extension/assets/55446273/6d32d57d-d99a-4c8c-81c9-d20337198a0d

cbadawi commented on src/app/features/asset-list/bitcoin/bitcoin-standards-loader/bitcoin-standards-loader.tsx:
```diff
@@ -0,0 +1,98 @@
+  filter?: AssetFilter;
+}
+
+export function BitcoinStandardsLoader({
```
went with the "list" suffix instead of "loader"

pete-watters commented on src/app/common/utils.ts:
```diff
@@ -292,3 +315,46 @@ export function removeTrailingNullCharacters(s: string) {
   return value.replace('-', '');
 }
+
+export function filterTokens<T>({
+  tokens,
+  filter = 'all',
+  getTokenIdentifier,
+}: {
+  tokens: T[];
+  filter: AssetFilter | Sip10CryptoAssetFilter;
+  getTokenIdentifier: (token: T) => string;
+}): T[] {
+  const accountIndex = useCurrentAccountIndex();
+  const allTokens = useAllTokens();
+
+  if (filter === 'all') return tokens;
+
+  return tokens.filter(token => {
+    const tokenIdentifier = getTokenIdentifier(token);
+    switch (filter) {
+      case 'supported':
+      case 'enabled':
+        return isTokenEnabled({ tokenIdentifier, allTokens, accountIndex });
+      case 'disabled':
+      case 'unsupported':
+        return !isTokenEnabled({ tokenIdentifier, allTokens, accountIndex });
+    }
+  });
```
Can you check this @cbadawi ?

cbadawi replied:
getTokenIdentifier should always be passed., will throw an error instead of console.error.

pete-watters commented on src/app/common/utils.ts:
```diff
@@ -20,6 +24,25 @@ interface MakeBitcoinTxExplorerLinkArgs {
   bitcoin: BitcoinChainConfig;
 }
 
+export type AssetFilter = 'all' | 'enabled' | 'disabled';
```
How about adding this new code in a separate file? In `src/app/common/filter-tokens.ts` or something like that? 

This `utils.ts` file is already really big and we should be gradually splitting it up and moving code out of it. That will also help us to gradually move functionality to our [mono-repo utils package](https://github.com/leather-io/mono/tree/dev/packages/utils).

You could also then maybe add a unit test to make sure that `isTokenEnabled` and `filterTokens` are working as expected

cbadawi replied:
fixed all the comments. just need to add the tests after work.

pete-watters commented on src/app/common/utils.ts:
```diff
@@ -292,3 +315,46 @@ export function removeTrailingNullCharacters(s: string) {
+  filter: AssetFilter | Sip10CryptoAssetFilter;
+  getTokenIdentifier: (token: T) => string;
+}): T[] {
+  const accountIndex = useCurrentAccountIndex();
```
This won't work as you are using React Hooks inside of a standard JS function. 

If you want to use hooks then `filterTokens` also needs to be a hook

Do you have`eslint` setup? To help you during development you can run `pnpm lint:eslint` to check for errors. 

![Screenshot 2024-07-02 at 12 49 14](https://github.com/leather-io/extension/assets/2938440/ed32a26f-54c8-4f5e-8592-52f90eee55a7)

cbadawi replied:
thanks was using had a really outdated eslint without realizing

pete-watters replied:
No problem, thanks for your contributions! 👏

pete-watters commented on src/app/components/crypto-asset-item/crypto-asset-item.layout.tsx:
```diff
@@ -33,39 +36,50 @@ export function CryptoAssetItemLayout({
 }: CryptoAssetItemLayoutProps) {
-  const { availableBalanceString, dataTestId, formattedBalance } =
-    parseCryptoAssetBalance(availableBalance);
+  let titleRight: ReactNode | undefined = undefined;
```
I don't think this is a good approach, using `let` and updating `titleRight` and `captionRight` .

If you use `const` it can help improve the maintainability and design of the code and reduce the complexity.

pete-watters commented on src/app/components/crypto-asset-item/crypto-asset-item.layout.utils.ts:
```diff
@@ -7,15 +7,19 @@ import { formatBalance } from '@app/common/format-balance';
 }
+
+export function getCryptoDataTestId(symbol: string) {
+  const dataTestId = CryptoAssetSelectors.CryptoAssetListItem.replace(
```
Here you could instead just `return`:

```
return  CryptoAssetSelectors.CryptoAssetListItem.replace(
    '{symbol}',
    symbol.toLowerCase()
  );
```

pete-watters commented on src/app/components/loaders/sip10-tokens-loader.tsx:
```diff
@@ -1,15 +1,41 @@
+  const getTokenIdentifier = (token: Token) => token.info.contractId;
+  const filteredTokens = filterTokens({ tokens, filter, getTokenIdentifier }) as Token[];
+
+  if (filter === 'all') {
+    // sort alex swappable tokens first for good ux
+    filteredTokens.sort((a, b) => {
+      const aPrincipal = getPrincipalFromContractId(getTokenIdentifier(a));
+      const bPrincipal = getPrincipalFromContractId(getTokenIdentifier(b));
+      return (
+        Number(swapAssets.some(swapAsset => swapAsset.principal === bPrincipal)) -
+        Number(swapAssets.some(swapAsset => swapAsset.principal === aPrincipal))
+      );
+    });
+  }
```
How about checking this @cbadawi

pete-watters commented on src/app/components/no-tokens-found/no-tokens-found.tsx:
```diff
@@ -0,0 +1,20 @@
+        className={css({
+          textAlign: 'center',
+          padding: '20px',
+          color: 'gray',
```
It would be better not to hard-code the color and font and to use our [tokens](https://github.com/leather-io/mono/blob/dev/packages/tokens/src/tokens.ts) or [typography](https://github.com/leather-io/mono/blob/dev/packages/tokens/src/typography.ts) if you can

pete-watters reviewed (changes requested):
Thanks for your hard work on this @cbadawi , there is some great work here and it's mostly good however there are some issues to address. 

I flagged some code style issues but the main thing that needs to be corrected first are to ensure there are no lint or typescript errors. 

- `pnpm lint:eslint` is flagging some mis-used react hooks 
- `pnpm typecheck` is finding some unused variables 

Once you have those issues resolved let me know and I can re-review and test

pete-watters commented:
> thanks @fbwoolf @fabric-8 , just noticed the color tokens & applied the changes. couple of minor issues though :
> 
> * could not make the color tokens work on the border and had to hardcode the colors. (works fine on backgroundColor property)
> * the chevron right icon has 2 chevrons. ">>"
> 
>  Screen.Recording.2024-07-02.at.1.21.43.PM.mov

There are two chevron icons, one with one `>` and one with two `>>` 

![Screenshot 2024-07-02 at 13 05 41](https://github.com/leather-io/extension/assets/2938440/9f4d1837-6c96-46eb-a093-a2195b9064ba)

fabric-8 commented:
@pete-watters Regarding the chevron icons (which seems resolved by now?): There are 2 different icons, the names are just really close. One is chevron**s**, the other one singular 😁

pete-watters commented:
> @pete-watters Regarding the chevron icons (which seems resolved by now?): There are 2 different icons, the names are just really close. One is chevron**s**, the other one singular 😁

Yes that is what I was saying here- https://github.com/leather-io/extension/pull/5568#issuecomment-2202994922

cbadawi commented:
@pete-watters thank you for your feedback. I applied the changes.

pete-watters commented:
> @pete-watters thank you for your feedback. I applied the changes.

Thanks @cbadawi , taking an in depth look now

pete-watters commented on src/app/components/loaders/runes-loader.tsx:
```diff
@@ -1,11 +1,25 @@
+  filter?: AssetFilter;
 }
-export function RunesLoader({ addresses, children }: RunesLoaderProps) {
+export function RunesLoader({ addresses, children, filter = 'all' }: RunesLoaderProps) {
```
@cbadawi I think this is a good suggestion

pete-watters commented on src/app/components/loaders/src20-tokens-loader.tsx:
```diff
@@ -9,8 +13,12 @@ export interface Src20TokenAssetDetails {
+export function Src20TokensLoader({ address, children, filter = 'all' }: Src20TokensLoaderProps) {
   const { data: tokens = [] } = useSrc20TokensByAddress(address);
-  return children(tokens);
+  const accountIndex = useCurrentAccountIndex();
```
This code seems almost the same as `Brc20TokensLoader`, do you think we could share it a bit better?

pete-watters commented on src/app/features/asset-list/bitcoin/src20-token-asset-list/src20-token-asset-list.tsx:
```diff
@@ -1,19 +1,29 @@
+  renderRightElement,
+  showBalance = true,
+}: Src20TokenAssetListProps) {
+  if (!tokens.length) return null;
```
@cbadawi can you please add this as I think it will throw an error if `tokens` is undefined

pete-watters commented on src/app/features/asset-list/stacks/sip10-token-asset-list/sip10-token-asset-list.tsx:
```diff
@@ -10,29 +12,37 @@ interface Sip10TokenAssetListProps {
           key={token.info.name}
           info={token.info}
           isLoading={isLoading}
-          marketData={priceAsMarketData(
-            getPrincipalFromContractId(token.info.contractId),
-            token.balance.availableBalance.symbol
-          )}
+          marketData={
+            showBalance
+              ? priceAsMarketData(
+                  getPrincipalFromContractId(token.info.contractId),
+                  token.balance.availableBalance.symbol
+                )
+              : null
+          }
           onSelectAsset={onSelectAsset}
+          renderRightElement={renderRightElement}
```
@cbadawi I think this is another good suggestion

pete-watters commented on src/app/features/asset-list/stacks/stacks-assets-list/stacks-assets-list.tsx:
```diff
@@ -0,0 +1,98 @@
+    >
+      {account => (
+        <>
+          {showStx && filter != 'unsupported' ? (
```
@cbadawi can you please refactor this to not use nested ternary statements? 

Nesting ternary statements makes the code hard to read and the logic tricky to follow

pete-watters commented on src/app/components/crypto-asset-item/crypto-asset-item.layout.tsx:
```diff
@@ -33,30 +37,37 @@ export function CryptoAssetItemLayout({
+    : { availableBalanceString: undefined, formattedBalance: undefined };
 
-  const titleRight = (
+  const titleRight: ReactNode | undefined = availableBalance ? (
```
@cbadawi can you please refactor this to not use nested ternary statements? 

Nesting ternary statements makes the code hard to read and the logic tricky to follow.

I'm also unsure about adding all of this code for optional props based on `availableBalance` or not

pete-watters commented on src/app/ui/components/toggle/toggle.tsx:
```diff
@@ -0,0 +1,88 @@
+      right: 0,
+      bottom: 0,
+      backgroundColor: 'ink.text-subdued',
+      border: '1px solid #F5F1ED',
```
Can you use our design tokens for colours, spacing, border etc. ? 

https://github.com/leather-io/mono/tree/dev/packages/tokens/src

cbadawi replied:
added for some where possible, but some specific pixel sizes dont exist such as 42px (width)
@pete-watters

pete-watters commented on src/app/features/manage-tokens/manage-tokens.tsx:
```diff
@@ -0,0 +1,97 @@
+      }
+    >
+      <Stack gap="space.05" pb="space.06" px="space.05" h="40vh">
+        {getManageTokenList()}
```
Maybe the lists can be just nested here instead of using a function?

pete-watters commented on src/app/features/asset-list/stacks/sip10-token-asset-list/sip10-token-asset-list.tsx:
```diff
@@ -24,15 +36,20 @@ export function Sip10TokenAssetList({
       {tokens.map(token => (
         <Sip10TokenAssetItem
-          balance={token.balance}
+          balance={showBalance ? token.balance : null}
```
Is the optional balance `showBalance` and `renderRightElement` so we can re-use `assetItem` to show a toggle or not?

Maybe it would be cleaner to have a new component just for showing the token name and the toggle? 
![Screenshot 2024-07-04 at 13 42 46](https://github.com/leather-io/extension/assets/2938440/3d9bdeb7-a7ae-42c8-985f-d5f8cadc476f)

cbadawi replied:
Current implementation was meant to be flexible; any right element, not just a toggle, could be used easily in the future instead of a balance. But you have a valid point of how having just a toggle element and a balance element is cleaner. Will do.

@pete-watters

kyranjamie commented on src/app/components/crypto-asset-item/crypto-asset-item.layout.tsx:
```diff
@@ -79,16 +89,15 @@ export function CryptoAssetItemLayout({
-    );
-
-  return <Box my="space.02">{content}</Box>;
+  return isInteractive ? (
+    <Pressable
+      data-testid={dataTestId}
+      onClick={() => onSelectAsset(symbol, contractId)}
+      my="space.02"
+    >
+      {content}
+    </Pressable>
+  ) : (
+    <Box my="space.02">{content}</Box>
+  );
```
Early return is a more readable pattern than this ternary, suggest reverting this change

kyranjamie commented on src/app/components/crypto-asset-item/crypto-asset-item.layout.tsx:
```diff
@@ -33,30 +37,37 @@ export function CryptoAssetItemLayout({
+  ) : undefined;
 
-  const captionRight = (
+  const captionRight: ReactNode | undefined = availableBalance ? (
```
Do we need this typing? Isn't it implicit?

kyranjamie reviewed (commented):
Thanks @cbadawi, just adding on with some suggestions to Pete's. 

Otherwise starting to look good.

cbadawi commented:
Did a minor refactor. Removed the need for the `showBalance`, `renderRightElement`. 

- Previous implementation was meant to be flexible, any component can be added as a right element. But i see how that bloated the code especially in the `CryptoAssetItemLayout` component.
- Current implementation has 2 right element variants to chose from `balance` & `toggle`. `rightElementVariant` defaults to `balance` and is responsible for choosing (in the `[token]-asset-lists.tsx` components) between the 2 components `CryptoAssetItemBalanceLayout` & `CryptoAssetItemBalanceLayout` (`CryptoAssetItemBalanceLayout` is equivalent to the now deleted `CryptoAssetItemLayout`).

Thank you both for your feedback.
@kyranjamie, @pete-watters apart from the spacing i couldnt find tokens for (mentioned above), i think the code is in pretty good shape. but feel free to give it the last nudge where needed. (especially that i'm just free on weekends)

markmhendrickson commented:
@314159265359879 to review a build after latest round of code review

pete-watters commented on src/app/common/filter-tokens.ts:
```diff
@@ -0,0 +1,91 @@
+  return tokenSetByUser?.enabled ?? isEnabledByDefault;
+}
+
+export function sortTokensBySwappability<T>({
```
There's a slight typo here for `Swappability`

pete-watters commented on src/app/features/asset-list/bitcoin/bitcoin-standards-list/bitcoin-standards-list.tsx:
```diff
@@ -0,0 +1,101 @@
+
+  const isReadOnly = variant === 'read-only';
+  return (
+    <>
```
I don't think this fragment is needed ? `<>` 🤔

pete-watters commented on src/app/features/asset-list/bitcoin/brc20-token-asset-list/brc20-token-asset-list.tsx:
```diff
@@ -22,11 +23,26 @@ interface Brc20TokenAssetDetails {
+    if (hasTokenSetter && tokens.length) hasTokenSetter(true);
+  }, [tokens.length, hasTokenSetter]);
+
+  if (!tokens.length) return null;
```
@cbadawi - there is another `if (!tokens.length) return null` on line 63 so lets get rid of one of them

pete-watters commented on src/app/features/asset-list/bitcoin/brc20-token-asset-list/brc20-token-asset-list.tsx:
```diff
@@ -22,14 +24,29 @@ interface Brc20TokenAssetDetails {
   const hasPositiveBtcBalanceForFees =
     variant === 'interactive' && balance.availableBalance.amount.isGreaterThan(0);
 
+  const icon = <Brc20AvatarIcon />;
```
I don't think this gives much benefit over just passing `<Brc20AvatarIcon />` directly

pete-watters commented on src/app/features/asset-list/bitcoin/runes-asset-list/runes-asset-list.tsx:
```diff
@@ -11,20 +15,49 @@ interface RuneTokenAssetDetails {
+        <CryptoAssetItemToggleLayout
+          captionLeft={captionLeft}
+          titleLeft={titleLeft}
+          key={`${symbol}`}
```
You can just use `key={symbol}` here instead but it could be good to also add the `index`

pete-watters commented on src/app/features/asset-list/stacks/stacks-assets-list/stacks-assets-list.tsx:
```diff
@@ -0,0 +1,128 @@
+  rightElementVariant: RightElementVariant;
+}
+
+function getStxCryptoAssetItem({
```
Can this be converted to a standalone component in its own file?

pete-watters commented:
> Did a minor refactor. Removed the need for the `showBalance`, `renderRightElement`.
> 
> * Previous implementation was meant to be flexible, any component can be added in the future as a right element simply, not just a toggle or a balance. But i see how that bloated the code especially in the `CryptoAssetItemLayout` component.
> * Current implementation is much less flexible but cleaner. We have only 2 right element variants to chose from : `balance` & `toggle`. `rightElementVariant` prop passed to the bitcoin & stacks lists defaults to `balance` and is responsible for choosing (in the `[token]-asset-lists.tsx` components) between the 2 components `CryptoAssetItemToggleLayout` & `CryptoAssetItemBalanceLayout` (`CryptoAssetItemBalanceLayout` is equivalent to the now deleted `CryptoAssetItemLayout`).
> 
> Thank you both for your feedback. @kyranjamie, @pete-watters apart from the spacing i couldnt find tokens for (mentioned above), i think the code is in pretty good shape. but feel free to give it the last nudge where needed. (especially that i'm just free on weekends)

Thanks @cbadawi . I re-reviewed and it's a lot better and pretty close now. I added a few suggestions, mostly small - a typo and some extra code. 

The AI bot did find this issue to fix though, [a double `tokens.length` check ](https://github.com/leather-io/extension/pull/5568/files#r1665404378)
`- if (!tokens.length) return null;` 

Other than that it's looking good and I tested it and it's working, great stuff

alter-eggo commented:
tested and src-20 token toggle is not working. 

also imo button text should be `Close` and not `Done`. or should be removed at all. as it just closes modal and doesn't have any other logic behind

![Screenshot 2024-07-09 at 02 46 43](https://github.com/leather-io/extension/assets/46521087/8cc39353-d8ba-4dff-8a94-6e295a768c86)

pete-watters commented:
> tested and src-20 token toggle is not working.
> 
> also imo button text should be `Close` and not `Done`. or should be removed at all. as it just closes modal and doesn't have any other logic behind
> 
> ![Screenshot 2024-07-09 at 02 46 43](https://private-user-images.githubusercontent.com/46521087/346743337-8cc39353-d8ba-4dff-8a94-6e295a768c86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MDE5OTMsIm5iZiI6MTcyMDUwMTY5MywicGF0aCI6Ii80NjUyMTA4Ny8zNDY3NDMzMzctOGNjMzkzNTMtZDhiYS00ZGZmLThhOTQtNmUyOTVhNzY4Yzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDA1MDgxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlZTNmMmMwMWQ1M2I2OWYxMjY2ODc3ZGYyMTBhZWI0MmMyNmE3NzdiMzBkZTYwYzFhODNhMTE5NDBlZmUzNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5NCCloCMzf4iuYHsD93UuXAGwcnN8L7SYkHRZdXD3Ys)

Good find with the SRC-20 token 👍 Maybe we don't even need any button there as we can just close the modal with the `X`

cbadawi commented:
@pete-watters  @alter-eggo  done 

https://github.com/leather-io/extension/assets/55446273/885b1e33-26c6-443a-bed9-1153a99adc6b

fabric-8 commented:
Good change with removing the "done" button, that was something that crossed my mind earlier today as well as not being necessary.

The toggle looked funky / square the other day when I checked, but looks like that's also fixed already?

cbadawi commented:
@fabric-8 yep all fixed

cbadawi commented:
any news here? @pete-watters feel free to take over / rebase / close this

pete-watters commented:
> any news here? @pete-watters feel free to take over / rebase / close this

Hey 👋 , 

It's on the teams list to help close it out but I have had to work on other things right now. It is on our radar and we want to help finish it so we will tackle it as soon as we can. Thanks again for all the hard work

314159265359879 commented:
I got some feedback from a user I thought was interesting and relates to this PR. 

They noticed that with the increased number of tokens that gets airdropped it also takes more time to load everything in their wallet. They'd like to see that after loading balances the selection of tokens to display prevents loading any additional data for tokens that are hidden. Basically do not load any metadata for tokens the user doesn't want to see anyway. 

Is that already how this is build? if not could that be scoped in?

markmhendrickson commented:
> Basically do not load any metadata for tokens the user doesn't want to see anyway.

This is a good idea. We can look at this when one of our dev turns their attention back to finalizing the PR. I've slated it for consideration during next week's sprint.
